File: sdks/python/aetherdb.py
"""
AetherDB Python Client SDK

Usage:
    from aetherdb import AetherDB
    db = AetherDB('https://your-aetherdb.workers.dev', api_key='your-key')
    doc = db.create_document('Hello world', metadata={'source': 'test'})
    results = db.search('Hello')
"""

from dataclasses import dataclass
from typing import Any, Optional
import json
import urllib.request
import urllib.error
import urllib.parse


@dataclass
class Document:
    id: str
    content: str
    metadata: dict
    created_at: str
    updated_at: str


@dataclass
class SearchHit:
    document_id: str
    score: float
    content_preview: str
    content: Optional[str] = None
    metadata: Optional[dict] = None


@dataclass
class HybridSearchHit:
    document_id: str
    score: float
    vector_score: float
    graph_score: float
    content_preview: str
    content: Optional[str] = None
    metadata: Optional[dict] = None


@dataclass
class FulltextHit:
    document_id: str
    snippet: str
    rank: float
    content: Optional[str] = None
    metadata: Optional[dict] = None


@dataclass
class Entity:
    id: str
    name: str
    entity_type: str
    document_id: str
    created_at: str


@dataclass
class Version:
    version_id: str
    content: str
    metadata: dict
    content_hash: str
    created_at: str


@dataclass
class HlcTimestamp:
    wall_ms: int
    counter: int


@dataclass
class Change:
    document_id: str
    content: str
    metadata: dict
    content_hash: str
    hlc_ts: HlcTimestamp
    origin_node: str
    deleted: bool


@dataclass
class SyncPullResponse:
    changes: list
    server_hlc: HlcTimestamp


@dataclass
class SyncPushResponse:
    accepted: int
    rejected: int
    server_hlc: HlcTimestamp


class AetherDBError(Exception):
    def __init__(self, status: int, message: str):
        self.status = status
        self.message = message
        super().__init__(f"AetherDB error {status}: {message}")


class AetherDB:
    def __init__(self, base_url: str, api_key: Optional[str] = None):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key

    def _headers(self) -> dict:
        h = {"Content-Type": "application/json"}
        if self.api_key:
            h["Authorization"] = f"Bearer {self.api_key}"
        return h

    def _request(self, method: str, path: str, body: Any = None) -> Any:
        url = f"{self.base_url}{path}"
        data = json.dumps(body).encode() if body else None
        req = urllib.request.Request(url, data=data, headers=self._headers(), method=method)
        try:
            with urllib.request.urlopen(req) as resp:
                if resp.status == 204:
                    return {}
                return json.loads(resp.read())
        except urllib.error.HTTPError as e:
            body_text = e.read().decode() if e.fp else ""
            try:
                err = json.loads(body_text)
                msg = err.get("error", body_text)
            except json.JSONDecodeError:
                msg = body_text
            raise AetherDBError(e.code, msg) from e

    # Documents
    def create_document(self, content: str, metadata: Optional[dict] = None, deduplicate: Optional[bool] = None) -> dict:
        body: dict = {"content": content, "metadata": metadata or {}}
        if deduplicate is not None:
            body["deduplicate"] = deduplicate
        return self._request("POST", "/v1/documents", body)

    def get_document(self, doc_id: str) -> Document:
        data = self._request("GET", f"/v1/documents/{doc_id}")
        return Document(**data)

    def update_document(self, doc_id: str, content: str, metadata: Optional[dict] = None) -> dict:
        return self._request("PUT", f"/v1/documents/{doc_id}", {"content": content, "metadata": metadata or {}})

    def delete_document(self, doc_id: str) -> None:
        self._request("DELETE", f"/v1/documents/{doc_id}")

    def get_versions(self, doc_id: str) -> list:
        data = self._request("GET", f"/v1/documents/{doc_id}/versions")
        return [Version(**v) for v in data.get("versions", [])]

    def list_documents(self, limit: int = 50, offset: int = 0, content_contains: Optional[str] = None,
                       created_after: Optional[int] = None, created_before: Optional[int] = None) -> dict:
        params = {"limit": str(limit), "offset": str(offset)}
        if content_contains:
            params["content_contains"] = content_contains
        if created_after is not None:
            params["created_after"] = str(created_after)
        if created_before is not None:
            params["created_before"] = str(created_before)
        qs = urllib.parse.urlencode(params)
        return self._request("GET", f"/v1/documents/list?{qs}")

    def bulk_import(self, documents: list, extract_entities: bool = False) -> dict:
        if extract_entities:
            return self._request("POST", "/v1/documents/bulk", {"documents": documents, "extract_entities": True})
        return self._request("POST", "/v1/documents/bulk", documents)

    # Search
    def search(self, query: str, top_k: int = 10, filter: Optional[dict] = None,
               return_content: bool = False) -> list:
        body: dict = {"query": query, "top_k": top_k, "return_content": return_content}
        if filter:
            body["filter"] = filter
        data = self._request("POST", "/v1/search/semantic", body)
        return [SearchHit(**h) for h in data.get("results", [])]

    def hybrid_search(self, query: str, top_k: int = 10, graph_depth: int = 2,
                      graph_weight: float = 0.3, filter: Optional[dict] = None,
                      return_content: bool = False) -> list:
        body: dict = {"query": query, "top_k": top_k, "graph_depth": graph_depth,
                      "graph_weight": graph_weight, "return_content": return_content}
        if filter:
            body["filter"] = filter
        data = self._request("POST", "/v1/search/hybrid", body)
        return [HybridSearchHit(**h) for h in data.get("results", [])]

    def fulltext_search(self, query: str, top_k: int = 10, return_content: bool = False) -> list:
        data = self._request("POST", "/v1/search/fulltext",
                             {"query": query, "top_k": top_k, "return_content": return_content})
        return [FulltextHit(**h) for h in data.get("results", [])]

    # SQL
    def query(self, sql: str) -> dict:
        return self._request("POST", "/v1/query", {"sql": sql})

    # Graph
    def get_entities(self, document_id: str) -> list:
        data = self._request("GET", f"/v1/graph/entities?document_id={document_id}")
        return [Entity(**e) for e in data.get("entities", [])]

    def get_related_documents(self, entity_name: str, depth: int = 2) -> list:
        data = self._request("GET", f"/v1/graph/entities/{urllib.parse.quote(entity_name)}/related?depth={depth}")
        return data.get("documents", [])

    # Sync
    def sync_pull(self, since: HlcTimestamp) -> SyncPullResponse:
        data = self._request("POST", "/v1/sync/pull", {"since": {"wall_ms": since.wall_ms, "counter": since.counter}})
        return SyncPullResponse(
            changes=data.get("changes", []),
            server_hlc=HlcTimestamp(**data.get("server_hlc", {"wall_ms": 0, "counter": 0})),
        )

    def sync_push(self, changes: list) -> SyncPushResponse:
        data = self._request("POST", "/v1/sync/push", {"changes": changes})
        return SyncPushResponse(
            accepted=data.get("accepted", 0),
            rejected=data.get("rejected", 0),
            server_hlc=HlcTimestamp(**data.get("server_hlc", {"wall_ms": 0, "counter": 0})),
        )

    # Health
    def health(self) -> dict:
        return self._request("GET", "/health")

File: sdks/typescript/aetherdb.ts
/**
 * AetherDB TypeScript Client SDK
 *
 * Usage:
 *   import { AetherDB } from './aetherdb';
 *   const db = new AetherDB('https://your-aetherdb.workers.dev', 'your-api-key');
 *   const doc = await db.createDocument('Hello world', { source: 'test' });
 *   const results = await db.search('Hello');
 */

export interface CreateDocumentResponse {
  id: string;
  created_at: string;
  embedded?: boolean;
  entities?: string;
  duplicate?: boolean;
}

export interface DocumentDto {
  id: string;
  content: string;
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface SearchHit {
  document_id: string;
  score: number;
  content_preview: string;
  content?: string;
  metadata?: Record<string, unknown>;
}

export interface SearchResponse {
  results: SearchHit[];
  count: number;
}

export interface HybridSearchHit {
  document_id: string;
  score: number;
  vector_score: number;
  graph_score: number;
  content_preview: string;
  content?: string;
  metadata?: Record<string, unknown>;
}

export interface HybridSearchResponse {
  results: HybridSearchHit[];
  count: number;
}

export interface FulltextHit {
  document_id: string;
  snippet: string;
  rank: number;
  content?: string;
  metadata?: Record<string, unknown>;
}

export interface FulltextResponse {
  results: FulltextHit[];
  count: number;
}

export interface EntityDto {
  id: string;
  name: string;
  entity_type: string;
  document_id: string;
  created_at: string;
}

export interface VersionDto {
  version_id: string;
  content: string;
  metadata: Record<string, unknown>;
  content_hash: string;
  created_at: string;
}

export interface ListDocumentsResponse {
  documents: DocumentDto[];
  total: number;
  limit: number;
  offset: number;
}

export interface BulkImportResponse {
  created: number;
  errors: number;
  embedded: number;
  entities_extraction: string;
  documents: Array<{ id: string; created_at: string }>;
  error_details: Array<{ index: number; error: string }>;
}

export interface HlcTimestamp {
  wall_ms: number;
  counter: number;
}

export interface Change {
  document_id: string;
  content: string;
  metadata: Record<string, unknown>;
  content_hash: string;
  hlc_ts: HlcTimestamp;
  origin_node: string;
  deleted: boolean;
}

export interface SyncPullResponse {
  changes: Change[];
  server_hlc: HlcTimestamp;
}

export interface SyncPushResponse {
  accepted: number;
  rejected: number;
  server_hlc: HlcTimestamp;
}

export interface HealthResponse {
  status: string;
  domain: string;
  documents: number;
  entities: number;
  relationships: number;
}

export class AetherDB {
  private baseUrl: string;
  private apiKey?: string;

  constructor(baseUrl: string, apiKey?: string) {
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.apiKey = apiKey;
  }

  private headers(): Record<string, string> {
    const h: Record<string, string> = { 'Content-Type': 'application/json' };
    if (this.apiKey) h['Authorization'] = `Bearer ${this.apiKey}`;
    return h;
  }

  private async request<T>(method: string, path: string, body?: unknown): Promise<T> {
    const resp = await fetch(`${this.baseUrl}${path}`, {
      method,
      headers: this.headers(),
      body: body ? JSON.stringify(body) : undefined,
    });
    if (!resp.ok) {
      const err = await resp.json().catch(() => ({ error: resp.statusText }));
      throw new Error(`AetherDB ${method} ${path}: ${resp.status} ${(err as any).error || resp.statusText}`);
    }
    if (resp.status === 204) return {} as T;
    return resp.json() as T;
  }

  // Documents
  async createDocument(content: string, metadata: Record<string, unknown> = {}, opts?: { deduplicate?: boolean }): Promise<CreateDocumentResponse> {
    return this.request('POST', '/v1/documents', { content, metadata, deduplicate: opts?.deduplicate });
  }

  async getDocument(id: string): Promise<DocumentDto> {
    return this.request('GET', `/v1/documents/${id}`);
  }

  async updateDocument(id: string, content: string, metadata: Record<string, unknown> = {}): Promise<CreateDocumentResponse> {
    return this.request('PUT', `/v1/documents/${id}`, { content, metadata });
  }

  async deleteDocument(id: string): Promise<void> {
    await this.request('DELETE', `/v1/documents/${id}`);
  }

  async getVersions(id: string): Promise<{ document_id: string; versions: VersionDto[]; count: number }> {
    return this.request('GET', `/v1/documents/${id}/versions`);
  }

  async listDocuments(opts?: { limit?: number; offset?: number; contentContains?: string; createdAfter?: number; createdBefore?: number }): Promise<ListDocumentsResponse> {
    const params = new URLSearchParams();
    if (opts?.limit) params.set('limit', String(opts.limit));
    if (opts?.offset) params.set('offset', String(opts.offset));
    if (opts?.contentContains) params.set('content_contains', opts.contentContains);
    if (opts?.createdAfter) params.set('created_after', String(opts.createdAfter));
    if (opts?.createdBefore) params.set('created_before', String(opts.createdBefore));
    return this.request('GET', `/v1/documents/list?${params}`);
  }

  async bulkImport(documents: Array<{ content: string; metadata?: Record<string, unknown> }>, opts?: { extractEntities?: boolean }): Promise<BulkImportResponse> {
    if (opts?.extractEntities) {
      return this.request('POST', '/v1/documents/bulk', { documents, extract_entities: true });
    }
    return this.request('POST', '/v1/documents/bulk', documents);
  }

  // Search
  async search(query: string, opts?: { topK?: number; filter?: Record<string, unknown>; returnContent?: boolean }): Promise<SearchResponse> {
    return this.request('POST', '/v1/search/semantic', {
      query, top_k: opts?.topK ?? 10,
      filter: opts?.filter, return_content: opts?.returnContent,
    });
  }

  async hybridSearch(query: string, opts?: { topK?: number; graphDepth?: number; graphWeight?: number; filter?: Record<string, unknown>; returnContent?: boolean }): Promise<HybridSearchResponse> {
    return this.request('POST', '/v1/search/hybrid', {
      query, top_k: opts?.topK ?? 10,
      graph_depth: opts?.graphDepth ?? 2,
      graph_weight: opts?.graphWeight ?? 0.3,
      filter: opts?.filter, return_content: opts?.returnContent,
    });
  }

  async fulltextSearch(query: string, opts?: { topK?: number; returnContent?: boolean }): Promise<FulltextResponse> {
    return this.request('POST', '/v1/search/fulltext', {
      query, top_k: opts?.topK ?? 10, return_content: opts?.returnContent,
    });
  }

  // SQL
  async query(sql: string): Promise<{ rows: Record<string, unknown>[]; count: number }> {
    return this.request('POST', '/v1/query', { sql });
  }

  // Graph
  async getEntities(documentId: string): Promise<{ entities: EntityDto[] }> {
    return this.request('GET', `/v1/graph/entities?document_id=${documentId}`);
  }

  async getRelatedDocuments(entityName: string, depth = 2): Promise<{ documents: Array<{ document_id: string; entity_name: string; hop_count: number }> }> {
    return this.request('GET', `/v1/graph/entities/${encodeURIComponent(entityName)}/related?depth=${depth}`);
  }

  // Sync
  async syncPull(since: HlcTimestamp): Promise<SyncPullResponse> {
    return this.request('POST', '/v1/sync/pull', { since });
  }

  async syncPush(changes: Change[]): Promise<SyncPushResponse> {
    return this.request('POST', '/v1/sync/push', { changes });
  }

  // Health
  async health(): Promise<HealthResponse> {
    return this.request('GET', '/health');
  }
}

export default AetherDB;
